File: samples/python/solve_constrained.py
import ags_solver
import math
import argparse
try:
    import sys
    sys.path.append('../3rd-party/global-optimization-test-problems/')
    from benchmark_tools.plot import plot_countours
    plot_found = True
except ImportError:
    plot_found = False

trials = []
def c1(x):
    global trials
    trials.append(x)
    return 0.01*(math.pow(x[0] - 2.2, 2) + math.pow(x[1] - 1.2, 2) - 2.25)
def c2(x):
    return 100 * (1 - math.pow(x[0] - 2, 2) / 1.44 - math.pow(0.5*x[1], 2))
def c3(x):
    return 10 * (x[1] - 1.5 - 1.5*math.sin(2*math.pi*(x[0] - 1.75)))
def obj_f(x):
    return -1.5*math.pow(x[0], 2) * math.exp(1 - pow(x[0], 2)
    - 20.25*math.pow(x[0] - x[1], 2)) - math.pow(0.5 * (x[1] - 1)*(x[0]- 1), 4) \
    * math.exp(2 - math.pow(0.5 * (x[0] - 1), 4) - math.pow(x[1] - 1, 4))

bounds = [[0, -1], [4, 3]]

def main(args):
    params = ags_solver.Parameters()
    params.eps = args.eps
    params.itersLimit = args.max_iters
    params.r = args.r
    params.evolventDensity = args.m
    params.epsR = args.epsR
    params.refineSolution = args.refine_loc

    solver = ags_solver.Solver()
    solver.SetParameters(params)
    solver.SetProblem([c1, c2, c3, obj_f], *bounds)
    point, val, idx = solver.Solve()

    calcCounters = solver.GetCalculationsStatistics()
    holderConstEstimations = solver.GetHolderConstantsEstimations()

    for i in range(len(calcCounters) - 1):
        print('Number of calculations of constraint # {}: {}'.format(i, calcCounters[i]))
    print('Number of calculations of objective: {}'.format(calcCounters[-1]))

    for i in range(len(holderConstEstimations) - 1):
        print('Estimation of Holder constant of function # {}: {}'.format(i, holderConstEstimations[i]))
    print('Estimation of Holder constant of objective: {}'.format(holderConstEstimations[-1]))

    if idx < 3:
        print('Feasible point not found')
    else:
        print('Optimal value: {}'.format(val))
        print('x = ' + str(point[0]) + ' y = ' + str(point[1]))

    if plot_found:
        evolvent_points = None
        if len(args.evolvent_points):
            evolvent_points = []
            with open(args.evolvent_points, 'r') as f:
                for line in f.readlines()[1:]:
                    coordinates = line.strip().split(',')
                    point = []
                    for coord in coordinates:
                        point.append(float(coord))
                    evolvent_points.append(point)
        plot_countours([c1, c2, c3, obj_f], bounds, 'example.pdf', points=trials, show=True, evolvent_points=evolvent_points)

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Sample for AGS solver')
    parser.add_argument('--m', type=int, default=12, help='Evolvent density')
    parser.add_argument('--r', type=float, default=3, help='Reliability parameter for the solver')
    parser.add_argument('--eps', type=float, default=0.01, help='Accuracy of the method')
    parser.add_argument('--epsR', type=float, default=0.1, help='eps-reserves for all constraints')
    parser.add_argument('--max_iters', type=int, default=10000, help='limit of iterations for the method')
    parser.add_argument('--refine_loc', action='store_true', help='Refine the global solution using a local optimizer')
    parser.add_argument('--evolvent_points', type=str, default='', help='')

    main(parser.parse_args())

File: samples/solve_constrained.cpp
#define _USE_MATH_DEFINES
#include <iostream>
#include <cmath>
#include <cmdline.h>

#include "solver.hpp"

using namespace ags;

void initParser(cmdline::parser& parser);

int main(int argc, char* argv[])
{
  cmdline::parser parser;
  initParser(parser);
  parser.parse_check(argc, argv);

  SolverParameters parameters;
  parameters.eps = parser.get<double>("accuracy");
  parameters.r = parser.get<double>("reliability");
  parameters.itersLimit = parser.get<int>("itersLimit");
  parameters.epsR = parser.get<double>("reserves");
  parameters.evolventDensity = parser.get<int>("evolventDensity");
  parameters.refineSolution = parser.exist("refineLoc");
  parameters.localMix = parser.get<int>("localMix");

  NLPSolver solver;
  solver.SetParameters(parameters);
  solver.SetProblem({
    [](const double* x) {return 0.01*(pow(x[0] - 2.2, 2) + pow(x[1] - 1.2, 2) - 2.25);},
    [](const double* x) {return 100 * (1 - pow(x[0] - 2, 2) / 1.44 - pow(0.5*x[1], 2));},
    [](const double* x) {return 10 * (x[1] - 1.5 - 1.5*sin(2*M_PI*(x[0] - 1.75)));},
    [](const double* x) {return -1.5*pow(x[0], 2) * exp(1 - pow(x[0], 2)
        - 20.25*pow(x[0] - x[1], 2)) - pow(0.5 * (x[1] - 1)*(x[0]- 1), 4)
        * exp(2 - pow(0.5 * (x[0] - 1), 4) - pow(x[1] - 1, 4));}
  }, {0, -1}, {4, 3});

  auto optimalPoint = solver.Solve();
  auto calcCounters = solver.GetCalculationsStatistics();
  auto holderConstEstimations = solver.GetHolderConstantsEstimations();

  for (size_t i = 0; i < calcCounters.size() - 1; i++)
    std::cout << "Number of calculations of constraint # " << i << ": " << calcCounters[i] << "\n";
  std::cout << "Number of calculations of objective: " << calcCounters.back() << "\n";

  for (size_t i = 0; i < holderConstEstimations.size() - 1; i++)
    std::cout << "Estimation of Holder constant of function # " << i << ": " << holderConstEstimations[i] << "\n";
  std::cout << "Estimation of Holder constant of objective: " << holderConstEstimations.back() << "\n";

  if (optimalPoint.idx < 3)
    std::cout << "Feasible point not found" << "\n";
  else
  {
    std::cout << "Optimal value: " << optimalPoint.g[optimalPoint.idx] << "\n";
    std::cout << "x = " << optimalPoint.y[0] << " y = " << optimalPoint.y[1] << "\n";
  }

  return 0;
}

void initParser(cmdline::parser& parser)
{
  parser.add<int>("evolventDensity", 'm', "", false, 12,
    cmdline::range(9, 16));
  parser.add<double>("reliability", 'r', "reliability parameter for the method",
    false, 3, cmdline::range(1., 1000.));
  parser.add<double>("accuracy", 'e', "accuracy of the method", false, 0.001);
  parser.add<double>("reserves", 'E', "eps-reserves for all constraints", false, 0.01);
  parser.add<int>("itersLimit", 'i', "limit of iterations for the method", false, 10000);
  parser.add("refineLoc", 'l', "Refine the global solution using a local optimizer");
  parser.add<int>("localMix", 'q', "local mix parameter", false, 0, cmdline::range(-20, 20));
}
